### PR TITLE
FYST-1122 Part 2: Update statefile tax year to 2024

### PR DESCRIPTION
**Pull request by arinchoi03**
## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/FYST-1122
## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- See the previous PR -- I was having incredibly hard time trying to stub out the TAX_YEAR constant that is being set by the `Rails.configuration` 
- Martha & I discussed how to fix and determined that we might as well update the tax year. First I renamed the 2023 folder for the fixtures to 2024 (so that they could be found correctly based on the current tax year) and then I will be moving the year folder (2024) out so that the structure is flatter & doesn't tie the personas to a tax year.
- Updated copy that was hardcoded to 2023 -> 2024 (some 2024-> 2025) but left NY/CTC copy as-is
## How to test?
- Describe the testing approach taken to verify the changes, including:
  - Unit/integration/manual tests
  - Test data used
- Specify any relevant testing environments used (e.g., development, staging, demo, Heroku).
- Risk Assessment
  - Risks or side effects associated with the changes and how they were mitigated.
  - Highlight areas that may need extra attention during code review or testing.
  - Paste SQL queries or output where relevant
## Screenshots (for visual changes)
- Before
- After

### CODE REVIEW

**Comment by arinchoi03** [comment]
at fbc72ad29f94d59188b8c7e5aee3d71b6eb7eb65 all specs were passing ✨

File: app/lib/submission_builder/document.rb [hunk 1]
```diff
@@ -18,6 +18,8 @@ def document
 
     def determine_default_schema_version_by_tax_year
       case @submission.tax_return&.year || @submission.data_source&.tax_return_year
+      when 2024 # TODO: update with 2024 schema when available
+        "2023v5.0"
```
Comment (arinchoi03) [inline app/lib/submission_builder/document.rb:22]: not sure where we can get this

File: spec/factories/az322_contributions.rb [hunk 2]
```diff
@@ -19,7 +19,7 @@
 #
 FactoryBot.define do
   factory :az322_contribution do
-    date_of_contribution_year { "2023" }
```
Comment (arinchoi03) [inline spec/factories/az322_contributions.rb:22]: update year dynamically using the rails config here

File: spec/factories/state_file_az_intakes.rb [hunk 3]
```diff
@@ -147,60 +147,60 @@
                state_file_az_intake: intake,
                charity_code: "22345",
                charity_name: "Heartland",
-               date_of_contribution: Date.parse("August 22 2023")
```
Comment (arinchoi03) [inline spec/factories/state_file_az_intakes.rb:150]: dynamically set the year

File: spec/features/state_file/complete_intake_spec.rb [hunk 4]
```diff
@@ -153,12 +153,6 @@
   end
 
   context "AZ", :flow_explorer_screenshot, js: true do
-    before do
-      # TODO: remove this stub when we update Rails.configuration.statefile_current_tax_year to 2024
-      allow(Rails.configuration).to receive(:statefile_current_tax_year).and_return 2024
-      stub_const('DateAccessible::TAX_YEAR', Date.new(2024))
-    end
-
```
Comment (arinchoi03) [inline spec/features/state_file/complete_intake_spec.rb:161]: this was the "fix" i was implementing to get the az feature specs to pass, which didn't work, so taking it out in this PR

File: spec/fixtures/state_file/fed_return_xmls/2024/az/old_sample.xml
Comment (arinchoi03) [inline spec/fixtures/state_file/fed_return_xmls/2024/az/old_sample.xml:1]: this file was added in the previous PR (that this branch is based off on)

File: spec/models/client_spec.rb [hunk 5]
```diff
@@ -535,7 +535,7 @@
       context "looking for CTC matches" do
         let(:display_type) { Intake::CtcIntake }
 
-        context "with a product year of 2023" do
```
Comment (arinchoi03) [inline spec/models/client_spec.rb:538]: should put these back

File: spec/models/tax_return_spec.rb [hunk 6]
```diff
@@ -932,7 +932,7 @@
       end
     end
 
-    context "when born on/after Jan 2, 1958 for tax year 2023" do
```
Comment (arinchoi03) [inline spec/models/tax_return_spec.rb:935]: maybe need to revisit if I need to make this dynamic

File: spec/support/helpers/state_file_intake_helper.rb [hunk 7]
```diff
@@ -11,7 +11,7 @@ def step_through_eligibility_screener(us_state:)
       choose "state_file_ny_eligibility_out_of_state_income_form_eligibility_part_year_nyc_resident_no"
       click_on "Continue"
 
-      expect(page).to have_text "In 2023, did you contribute to a 529 college savings account, or did you withdraw funds from a 529 account and use them for non-qualified expenses?"
```
Comment (arinchoi03) [inline spec/support/helpers/state_file_intake_helper.rb:14]: update to using config here

File: spec/features/state_file/complete_intake_spec.rb [hunk 8]
```diff
@@ -306,24 +302,24 @@
...
+      expect(page.body).to include('<QualParentsAncestors>')
+      expect(page.body).to include('<WageAmIndian>100</WageAmIndian>')
+      expect(page.body).to include('<CompNtnlGrdArmdFrcs>100</CompNtnlGrdArmdFrcs>')
+
+      assert_flow_explorer_sample_params_includes_everything('az')
+
+      perform_enqueued_jobs
+      submission = EfileSubmission.last
+      expect(submission.submission_bundle).to be_present
+      expect(submission.current_state).to eq("queued")
```
Comment (arinchoi03) [inline spec/features/state_file/complete_intake_spec.rb:322]: this passes now ✨  @DrewProebstel

File: app/controllers/state_file/questions/df_return_sample.xml [hunk 9]
```diff
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<Return xmlns="http://www.irs.gov/efile" xmlns:efile="http://www.irs.gov/efile" returnVersion="2023v5.0">
+<Return xmlns="http://www.irs.gov/efile" xmlns:efile="http://www.irs.gov/efile" returnVersion="2023v2.0">
```
Comment (jenny-heath) [inline app/controllers/state_file/questions/df_return_sample.xml:2]: [dust] i don't think i remember or can find any place where we actually use `returnVersion` from the df fixtures. is there a reason to change this?

File: app/lib/submission_builder/document.rb (hunk 1 above)
Comment (jenny-heath) [reply app/lib/submission_builder/document.rb:22]: i tried to look for all usages of `@schema_version` in children of this class and as far as i can tell it's only used for federal stuff? so we should be good

Comment (jenny-heath) [reply app/lib/submission_builder/document.rb:22]: (federal stuff = building federal returns i.e. CTC code)

File: app/views/state_file/landing_page/edit.html.erb [hunk 10]
```diff
@@ -65,7 +65,7 @@
               <% end %>
             <% else %>
               <div class="white-group spacing-below-60">
-                <%= t(".#{@state_code}.help_text_html", default: t(".help_text_html")) %>
+                <%= t(".#{@state_code}.help_text_html", default: t(".help_text_html"), filing_year: Rails.configuration.statefile_current_tax_year) %>
```
Comment (jenny-heath) [inline app/views/state_file/landing_page/edit.html.erb:68]: very minor nit: it looks like everywhere else this is set to an instance variable? personally i don't have a preference between the two but we should probably stick to one

File: app/controllers/state_file/questions/az_eligibility_out_of_state_income_controller.rb [hunk 11]
```diff
@@ -2,6 +2,8 @@ module StateFile
   module Questions
     class AzEligibilityOutOfStateIncomeController < QuestionsController
       include EligibilityOffboardingConcern
+
+      before_action -> { @filing_year = Rails.configuration.statefile_current_tax_year }
```
Comment (jenny-heath) [inline app/controllers/state_file/questions/az_eligibility_out_of_state_income_controller.rb:6]: so admittedly i forgot about this until part way into reviewing this PR but if you go look at `StateFileControllerConcern` there is a helper method called `current_tax_year` which actually gives you access to this config variable via the `MultiTenantService`. so i think you should just remove all of these new before_actions and use that helper method in the views instead

File: config/application.rb [hunk 12]
```diff
@@ -83,7 +83,7 @@ class Application < Rails::Application
     config.middleware.use Middleware::RejectBadlyEncodedHeaders
     config.gyr_current_tax_year = 2023
     config.ctc_current_tax_year = 2021
-    config.statefile_current_tax_year = 2023
+    config.statefile_current_tax_year = 2024
```
Comment (jenny-heath) [inline config/application.rb:86]: 🥳

File: config/locales/en.yml [hunk 13]
```diff
@@ -2361,21 +2361,21 @@ en:
         edit:
           body_html: You cannot use this tool this year because you %{ineligible_reason}. <strong>Is this a mistake?</strong> <a href=%{prev_link}>Go back to correct.</a>
           ineligible_reason:
-            eligibility_529_for_non_qual_expense: withdrew money from a 529 college savings account and used it for a non-qualifying expense in 2023
+            eligibility_529_for_non_qual_expense: withdrew money from a 529 college savings account and used it for a non-qualifying expense in 2024
```
Comment (jenny-heath) [inline config/locales/en.yml:2364]: is there a reason some of these are hard-coded? at first i thought it was just NY stuff but this is AZ

File: config/locales/en.yml [hunk 14]
```diff
@@ -2361,21 +2361,21 @@ en:
...
             eligibility_emergency_rental_assistance: received emergency rental assistance in 2024
             eligibility_home_different_areas: or your spouse had your permanent homes in two different counties, cities, towns, or taxing areas on the last day the tax year
             eligibility_homebuyer_withdrawal: withdrew money from a first-time homebuyer account and used it for anything other than purchasing a home
             eligibility_homebuyer_withdrawal_mfj: or your spouse withdrew money from a first-time homebuyer account and used it for anything other than purchasing a home
-            eligibility_lived_in_state: did not live in %{state} for all of 2023
+            eligibility_lived_in_state: did not live in %{state} for all of 2024
```
Comment (jenny-heath) [inline config/locales/en.yml:2369]: same question

File: spec/controllers/state_file/questions/az_public_school_contributions_controller_spec.rb [hunk 15]
```diff
@@ -141,7 +141,7 @@
           amount: 100,
           date_of_contribution_month: '8',
           date_of_contribution_day: "12",
-          date_of_contribution_year: "2023"
+          date_of_contribution_year: "2024"
```
Comment (jenny-heath) [inline spec/controllers/state_file/questions/az_public_school_contributions_controller_spec.rb:144]: looks like for other similar specs you used the config?

File: spec/features/state_file/complete_intake_spec.rb [hunk 16]
```diff
@@ -47,15 +49,15 @@
       end
       click_on I18n.t("general.continue")
 
-      expect(page).to have_text I18n.t("state_file.questions.nyc_residency.edit.title", year: 2023)
-      choose "I did not live in New York City at all in 2023"
+      expect(page).to have_text I18n.t("state_file.questions.nyc_residency.edit.title", year: 2024)
+      choose "I did not live in New York City at all in 2024"
```
Comment (jenny-heath) [inline spec/features/state_file/complete_intake_spec.rb:53]: should this be `year: filing_year` etc?

File: spec/features/state_file/edit_return_spec.rb
Comment (jenny-heath) [inline spec/features/state_file/edit_return_spec.rb:1]: more hard-coded `2024`s?

File: spec/lib/submission_builder/ty2022/states/az/az_return_xml_spec.rb [hunk 17]
```diff
@@ -156,7 +156,7 @@
         expect(xml.css('CharityInfo').count).to eq 4
         expect(xml.css('ContinuationPages').count).to eq 1
 
-        expect(xml.at('CharityInfo QualCharityContrDate').text).to eq '2023-08-22'
+        expect(xml.at('CharityInfo QualCharityContrDate').text).to eq '2024-08-22'
```
Comment (jenny-heath) [inline spec/lib/submission_builder/ty2022/states/az/az_return_xml_spec.rb:159]: probably another opportunity to set the year dynamically?

File: spec/lib/submission_builder/ty2024/states/nc/documents/d400_spec.rb [hunk 18]
```diff
@@ -126,12 +126,12 @@
     context "qw filers" do
       let(:intake) { create(:state_file_nc_intake, filing_status: "qualifying_widow") }
       before do
-        intake.direct_file_data.spouse_date_of_death = "2023-09-30"
+        intake.direct_file_data.spouse_date_of_death = "2024-09-30"
```
Comment (jenny-heath) [inline spec/lib/submission_builder/ty2024/states/nc/documents/d400_spec.rb:129]: same

File: spec/models/direct_file_data_spec.rb [hunk 19]
```diff
@@ -5,7 +5,7 @@
   let(:direct_file_data) { DirectFileData.new(xml.to_s) }
 
   [
-    ["tax_return_year", 2023],
+    ["tax_return_year", 2024],
```
Comment (jenny-heath) [inline spec/models/direct_file_data_spec.rb:8]: same

File: spec/models/state_file_az_intake_spec.rb [hunk 20]
```diff
@@ -100,7 +100,7 @@
                routing_number: "123456789",
                account_number: "123",
                withdraw_amount: 123,
-               date_electronic_withdrawal: Date.parse("April 1, 2023")
+               date_electronic_withdrawal: Date.parse("April 1, 2024")
```
Comment (jenny-heath) [inline spec/models/state_file_az_intake_spec.rb:103]: same

File: spec/models/state_file_md_intake_spec.rb [hunk 21]
```diff
@@ -130,7 +130,7 @@
                routing_number: "123456789",
                account_number: "123",
                withdraw_amount: 123,
-               date_electronic_withdrawal: Date.parse("April 1, 2023"),
+               date_electronic_withdrawal: Date.parse("April 1, 2024"),
```
Comment (jenny-heath) [inline spec/models/state_file_md_intake_spec.rb:133]: same

File: spec/personas/persona_spec.rb [hunk 22]
```diff
@@ -1,8 +1,8 @@
 require "rails_helper"
 
 describe "personas" do
-  context "2023" do
-    let(:tax_year) { 2023 }
+  context "2024" do
```
Comment (jenny-heath) [inline spec/personas/persona_spec.rb:4]: same

File: spec/models/tax_return_spec.rb (hunk 6 above)
Comment (jenny-heath) [reply spec/models/tax_return_spec.rb:935]: we don't have tax returns for state file so i think we can leave this file alone actually

File: spec/features/state_file/complete_intake_spec.rb (hunk 16 above)
Comment (jenny-heath) [reply spec/features/state_file/complete_intake_spec.rb:53]: oh i guess on closer inspection it's just NY so i guess i don't care but it does feel weird that we're making tons of updates to NY stuff when we should in theory be deleting it * sigh *

**Review by jenny-heath — APPROVED** [review]
looks mostly good! the only changes i would request are:
- use the existing helper method to get the current filing year in view (see comment)
- there are some places in tests and in translations where the year is still hard-coded but could be dynamic? maybe there's a reason for these i'm not seeing. if so, happy to chat about it!

File: spec/features/state_file/complete_intake_spec.rb (hunk 16 above)
Comment (arinchoi03) [reply spec/features/state_file/complete_intake_spec.rb:53]: i just updated them to be consistent 🤷

File: config/locales/en.yml (hunk 13 above)
Comment (arinchoi03) [reply config/locales/en.yml:2364]: good catch! yes ill update

**Comment by arinchoi03** [comment]
going to try to unified approach to getting filing_year in the views from `current_tax_year` if available

**Pull request merged**